pacsman/filesystem_dev_client.py

```diff
@@ -48,78 +75,162 @@ def __init__(self, dicom_dir: str, dicom_source_dir: str, *args, **kwargs) -> No
         os.makedirs(self.dicom_dir, exist_ok=True)
         self.dicom_source_dir = dicom_source_dir
 
-        self.dicom_datasets: Dict[str, Dataset] = {}
+        self.cached_dicom_datasets: Dict[str, Dataset] = {}
 
-        for dicom_file in glob.glob(f'{dicom_source_dir}/**/*.dcm', recursive=True):
-            self._read_and_add_data_set(dicom_file)
+        # load and use the index if it is present and its hash matches the current dir
+        index_path = self._filepath(INDEX_FILENAME)
+        if os.path.exists(index_path):
+            with open(index_path, 'rb') as f:
+                self.index = pickle.load(f)
+                if self.index.dicom_source_dir_hash != self._dicom_source_dir_hash():
+                    self._reindex_source_dir()
+        else:
+            self._reindex_source_dir()
 
-    def _read_and_add_data_set(self, filename: str) -> None:
-        filepath = self._filepath(filename)
-        self._add_dataset(dcmread(filepath, stop_before_pixels=True), filepath)
+    def _filepath(self, filename: str) -> str:
+        return os.path.join(self.dicom_source_dir, filename)
 
-    def _add_dataset(self, dataset: Dataset, filepath: str = None) -> None:
-        if filepath is None:
-            filepath = self._filepath(dicom_filename(dataset))
-        self.dicom_datasets[filepath] = dataset
+    def _dicom_source_dir_hash(self):
+        """
+        Build a hash of DICOM files in the `dicom_source_dir` using their names, sizes,
+        and modification times.
+        """
+        h = hashlib.md5()
+        for dicom_file in glob.glob(f'{self.dicom_source_dir}/**/*.dcm', recursive=True):
```

Comment:
It seems like we are assuming all dicom files end in `.dcm`.  I think this is fine for now, but perhaps we should add a comment about it?